lib/internal/modules/esm/loader.js

```diff
@@ -179,39 +178,37 @@ function nextHookFactory(chain, meta, { validateArgs, validateOutput }) {
  * the main module and everything in its dependency graph.
  */
 class ESMLoader {
-  /**
-   * Prior to ESM loading. These are called once before any modules are started.
-   * @private
-   * @property {KeyedHook[]} globalPreloaders Last-in-first-out
-   *  list of preload hooks.
-   */
-  #globalPreloaders = [];
-
-  /**
-   * Phase 2 of 2 in ESM loading.
-   * @private
-   * @property {KeyedHook[]} loaders Last-in-first-out
-   *  collection of loader hooks.
-   */
-  #loaders = [
-    {
-      fn: defaultLoad,
-      url: 'node:internal/modules/esm/load',
-    },
-  ];
-
-  /**
-   * Phase 1 of 2 in ESM loading.
-   * @private
-   * @property {KeyedHook[]} resolvers Last-in-first-out
-   *  collection of resolver hooks.
-   */
-  #resolvers = [
-    {
-      fn: defaultResolve,
-      url: 'node:internal/modules/esm/resolve',
-    },
-  ];
+  #hooks = {
+    __proto__: null,
```

Comment:
Do we need a null-prototype here?
```suggestion
```

Reply:
I'm pretty sure you originally requested `ObjectCreate(null)`.

I think your rationale was because we destructure from these (it's changed slightly, but I think your original concern is still applicable/valid), making it vulnerable to prototype pollution.

Reply:
V8 optimization process is not straightforward, so it’s very possible that using null-prototype is not an issue, but if we know the shape of the object, we should chose the fast path, and make sure we don’t use it unsafely.

Destructuring has nothing inherently unsafe, and has nothing to do with prototype pollution, so I think there’s been a misunderstanding.

Reply:
Ahh, duh, it's just when we haven't explicitly set the properties that it's unsafe.